job-info: add new job-info.list-inactive service

chu11 opened the pull request:
Per discussion on #2683, this allows callers to get a subset of the inactive jobs in the job info module.  The subset is limited via a timestamp parameter, which allows callers to only get jobs newer than an inputted timestamp (i.e. job t_inactive > inputted timestamp).

This is quite specific for the issue problem and not very generic.  I went back and forth on making it a bit more generic, but eventually came to this very specific implementation.  One of the major reasons is b/c the pending jobs list is sorted in priority order, so making a generic "limit based on timestamp" service would have to scan the entire pending jobs list.

This is built on top of issue #2720.  It could be spliced out, but given the likely conflicts, just thought it'd be easier to build on top of it.

chu11 commented:
just throwing up an example:

Here's all the inactive jobs & their inactive timestamp

```
>flux job list-inactive | jq ".id, .t_inactive"
122926661632
1581555364.3478162
119990648832
1581555364.1802206
115762790400
1581555363.9248183
106384326656
1581555363.3657677
101183389696
1581555363.0552006
94791270400
1581555362.6729515
86117449728
1581555362.1650386
```

lets just get the job ids after a certain timestamp

```
>flux job list-inactive --timestamp 1581555363.3657677 | jq ".id"
122926661632
119990648832
115762790400
```

chu11 commented:
re-pushed with a couple EPROTO/EINVAL tests, hopefully up the coverage a tad

chu11 commented:
> (I may have messed up these new tests, will follow up :-)

Well, to my surprise

`jq -c -n "{timestamp:0.0}"`

leads to

`{"timestamp":0}`

which when parsing with `jansson _unpack()` and `s:f`, leads to a protocol error, b/c its an integer and not a real.

`jq` says this is ok and legal json.  https://github.com/stedolan/jq/issues/356

unsure why jansson can't read an integer into a real.

Temporarily fixed by just putting some real numbers in instead of "0.0".

chu11 commented:
Rebased now that #2720 is in.

grondo commented:
> unsure why jansson can't read an integer into a real.

I think if you used the format `s:F` instead of `s:f` jansson would unpack an integer to real. From [the docs](https://jansson.readthedocs.io/en/2.8/apiref.html#parsing-and-validating-values:)

> `f` (real) [double]
>     Convert a JSON real to C double.
> `F` (integer or real) [double]
>     Convert a JSON number (integer or real) to C double.

The thinking might be if your protocol *requires* a JSON real, then use `f`, if you only require a number then use `F`.

Maybe in this PR we want `F` not `f`?

grondo commented:
My other comment for the new `job-info.list-inactive` rpc is that it doesn't have a built in count like `job-info.list`, which would protect the service from accidental DoS by someone doing something like
```
$ flux job list-inactive --timestamp 0.0
``` 

when there are millions of jobs.

For now, it might be sufficient to set a default limit of 1000 jobs returned, overridden on the cmdline and the request message like `job-info.list`. In the future, or if you are feeling ambitious, a "more data available" flag could be added to the response, along with a `restart` timestamp in the request, which would allow a caller to fetch a large list of jobs in chunks.

chu11 commented:
> f (real) [double]
Convert a JSON real to C double.
F (integer or real) [double]
Convert a JSON number (integer or real) to C double.

Gah! I was looking at the documentation just now and realized that I had scrolled to the "json_pack()" area, thus missing the `F`.

chu11 commented:
rebased & re-pushed with adjusted listed above:

- `f` to `F` in the `job-info` module
- added a `max_entries` field just like `job-info.list`, with option in `flux job` and tests.

chu11 commented:
> Looks good! Want this merged for the v0.16 tag?

Lets do it!
